### PR TITLE
Use under_13 to determine Dance Party Age - Clean-up

### PR DESCRIPTION
**Pull request by epeach**
https://codedotorg.atlassian.net/browse/STAR-2159
Summary - Teachers were reporting that <13, logged-in students were able to access pg-13 songs.

Couldn't reproduce locally, but it seems like we were having an issue differentiating 'student' and 'student_y' and were defaulting to turning the filterOff. I added a check for whether the currentUser is a student and has under_13 set to false to determine whether to turn the filterOff.

If you were able to reproduce the problem locally, please pull and test locally!

What's weird is that our tests, where we check if a under13 student can't access the filtered songs was still working (age_filter.feature), so this definitely seems like something has changed in how we register students as over/under 13. Worth looking into to further understand and eliminate potential future regressions, but my hope is that this alleviates the current need.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by breville** [comment]
After assuming identify of an affected user on production and hitting some relevant breakpoints in the code, I think I know the root cause and have an idea to add to this fix.

We actually set the `userType` in redux twice:

- via `setUserType` [here](https://github.com/code-dot-org/code-dot-org/blob/6c3dfe23b6ec58b71f49eb23c8ec40330a58b990/apps/src/templates/currentUserRedux.js#L63).  This value comes from the DOM.  This was setting `"student_y"` for a < 13 student.
- via `setInitialData` [here](https://github.com/code-dot-org/code-dot-org/blob/6c3dfe23b6ec58b71f49eb23c8ec40330a58b990/apps/src/templates/currentUserRedux.js#L73).  This value comes from an async request to https://studio.code.org/api/v1/users/current.  This was setting `"student"` for a < 13 student.

It seems that this is a race condition, which would explain why a localhost repro was elusive.  On localhost, we hit `setInitialData` then `setUserType`.  On production, we hit `setUserType` then `setInitialData`, overwriting the `"student_y"` with `"student"`.

To fix, we could update https://studio.code.org/api/v1/users/current to also return `"student_y"`, but that value was really just invented for a cookie that could be used by pegasus in [this](https://github.com/code-dot-org/code-dot-org/pull/22043) PR.  The value `"student_y"` doesn't really exist anywhere else in the dashboard.  Presumably there would be unintended side-effects on other code to change this behaviour.

So I think the solution here is actually a good one.  But additionally, we should revisit the original approach to setting `usertype` [here](https://github.com/code-dot-org/code-dot-org/pull/25724/files#diff-a25e05619ab807078b355d8a7de339a451e0acfcf114666d35078e219c757d22R39-R40) in [this](https://github.com/code-dot-org/code-dot-org/pull/25724) PR, so that we just set the `user_type` from `current_user`, which should just be `"teacher"` or `"student"` but never `"student_y"`.  Also, should we send `under13` through with the `setUserType`?

**Comment by epeach** [comment]
> To fix, we could update https://studio.code.org/api/v1/users/current to also return "student_y", but that value was really just invented for a cookie that could be used by pegasus in https://github.com/code-dot-org/code-dot-org/pull/22043 PR. The value "student_y" doesn't really exist anywhere else in the dashboard. Presumably there would be unintended side-effects on other code to change this behaviour.

It seems like with the fragility of the student_y system, I'm partial to moving away from that and towards the `under13` system that I use here - it seems pretty established in the rest of our code.

> But additionally, we should revisit the original approach to setting usertype [here](https://github.com/code-dot-org/code-dot-org/pull/25724/files#diff-a25e05619ab807078b355d8a7de339a451e0acfcf114666d35078e219c757d22R39-R40) in https://github.com/code-dot-org/code-dot-org/pull/25724 PR, so that we just set the user_type from current_user, which should just be "teacher" or "student" but never "student_y". Also, should we send under13 through with the setUserType?

To check my understanding, is the suggestion that we remove/clean-up the code in the linked PR that refers to the userType and setting the userType?

**Comment by breville** [comment]
> To check my understanding, is the suggestion that we remove/clean-up the code in the linked PR that refers to the userType and setting the userType?

The suggestions, all for this PR:

- `application.html.haml` gets its `usertype` from `current_user.user_type` and passes that through, rather than from the cookie.  (This prevents `"student_y"` from being passed through.)
- For both approaches, set the new < 13 value at the same time we set the user type.  This way we update the client's redux store more consistently.

**Comment by epeach** [comment]
@breville PTAL - thank you! I updated application.haml.html, but I went with a different direction for setUserType. We were setting the scriptdata in application from current_user.user_type and then in code_studio.js taking the script data to set current_user.user_type. Since the setUserType was only being hit from code_studio.js and that was added specifically for dance party filtering, I figured we could just get rid of that whole path. Now currentUserRedux is the source of truth for userType and we only set userType from setInitialData. Let me know your thoughts on this route. Cheers!

**Comment by breville** [comment]
> Now currentUserRedux is the source of truth for userType and we only set userType from setInitialData.

I like the idea of simplifying here, but I'm not sure whether we can count on the value set by `setInitialData` before the dance party begins.  I don't know much about this code, but it appears to be some header-related code that initiates the retrieval and then sets this data [here](https://github.com/code-dot-org/code-dot-org/blob/2a118c192a8ba1e92f0ddfea3a29dc71f9bb1914/apps/src/code-studio/header.js#L200).  

To test this hunch, I blocked the URL for this async call in the Chrome debugger.  The dance party still loaded, but `setInitialData` was never called, and the user type was coming via `setUserType` instead.  Please double-check my theory, but I think that if this API was particularly slow to return, or never did, then we would not be able to count on a value coming from it.

**Comment by epeach** [comment]
I'll definitely look into that. One thought is that I would like to separate this PR into two PRs now that we are getting deeper into clean-up code and the high-pri bug is still live for users. I believe we should be able to merge my first commit to immediately to solve the problem our users are seeing with no regression. Then we can put these subsequent changes into a PR where we have the time to fully explore the clean-up work.

**Comment by epeach** [comment]
Broke out the PR as mentioned above. https://github.com/code-dot-org/code-dot-org/pull/45343
Even if we end up going in a different direction in the long run, this will alleviate the current error that we received Zendesk tickets about.

**Comment by epeach** [comment]
> To test this hunch, I blocked the URL for this async call in the Chrome debugger. The dance party still loaded, but `setInitialData` was never called, and the user type was coming via `setUserType` instead.

When `setInitialData` isn't called, we fall back to filter on, which is expected as we cannot determine whether they can view Pg13 material. So this doesn't seem like the worst case. `Header.js` is loaded on each page and calls 'setInitialData', so if it fails on one page load, the next one has a chance of working and getting us in a good state.

If the concern is that we have a race condition between `DanceVisualizationColumn` looking at the `currentUserRedux` and the `header.js` setting the `currentUserRedux`, I'd rather us look at how we are managing `signedInState` in `currentUserRedux` and try to piggyback on those calls.

**Comment by breville** [comment]
I strongly feel that we should use the DOM to pass the user type & < 13 status for now.  We should never rely on being fortunate with timings, and we should never be happy with fallback behaviour when we don't need to be, because that would mean we have to support multiple unpredictable code paths when we only need one.

As for whether a tutorial can rely on using a value retrieved by the header, that's a deeper architectural question that I think is a bit out of scope right now.  Currently it's clear that a tutorial should not rely on the initial value set by `setInitialData`, but that if it does rely upon that value, it should also cope with that value changing later.  It seems problematic that things work this way, but I think for now we can at least make both  `setUserType` and `setInitialData` work consistently.

File: apps/src/templates/currentUserRedux.js [hunk 1]
```diff
@@ -33,7 +37,8 @@ const initialState = {
   userName: null,
   userType: 'unknown',
   signInState: SignInState.Unknown,
-  hasSeenStandardsReportInfo: false
+  hasSeenStandardsReportInfo: false,
+  under13: true
```
Comment (breville) [inline apps/src/templates/currentUserRedux.js:42]: This default makes sense for our scenario, and likely others, but might be worth adding a comment here explaining why we default to `true`.

**Comment by breville** [comment]
After a successful pairing session, just a quick note that we discovered that `current_user` is not available for most `dance-2019` levels because they are [cached](https://github.com/code-dot-org/code-dot-org/blob/d83b32c1665c9495e710757a5a198db26d48dd0f/cookbooks/cdo-varnish/libraries/http_cache.rb#L43), so we will rely on the cookie after all, but will have `application.html.haml` transform a `"student_y"` cookie into the constituent parts (`user_type: "student"` and `under_13: true`) that we would have expected from the server.  This avoids dance-specific code from having to differentiate between values in the redux store being set via either `setUserType` or `setInitialData`.

**Comment by epeach** [comment]
@breville - PTAL! I added the changes we worked on while pairing and added some comments.

File: dashboard/app/views/layouts/application.html.haml [hunk 2]
```diff
@@ -62,8 +62,16 @@
     - if current_user && current_user.age.nil?
       = render partial: 'layouts/age_interstitial'
 
+
     - cookie_key = environment_specific_cookie_name('_user_type')
     - user_type = request.cookies[cookie_key]
+    -# The _user_type cookie is set in devise.rb to differentiate under13 students as 'student_y'
+    -# Splitting that single value here into two variables to align with user_type standards of "teacher/student"
+    - if user_type ==  == 'student_y'
```
Comment (breville) [inline dashboard/app/views/layouts/application.html.haml]: typo with double `==`

Comment (epeach) [reply dashboard/app/views/layouts/application.html.haml]: Oof. Good catch. Just pushed a fix.

File: dashboard/app/views/layouts/application.html.haml [hunk 3]
```diff
@@ -62,8 +62,16 @@
     - if current_user && current_user.age.nil?
       = render partial: 'layouts/age_interstitial'
 
+
     - cookie_key = environment_specific_cookie_name('_user_type')
     - user_type = request.cookies[cookie_key]
+    -# The _user_type cookie is set in devise.rb to differentiate under13 students as 'student_y'
+    -# Splitting that single value here into two variables to align with user_type standards of "teacher/student"
```
Comment (breville) [inline dashboard/app/views/layouts/application.html.haml]: nit: Add `.`s for ends of sentences here and in other comments too.

File: apps/src/dance/DanceVisualizationColumn.jsx [hunk 4]
```diff
@@ -83,7 +84,8 @@ class DanceVisualizationColumn extends React.Component {
     // userType - 'teacher', assumed age > 13. 'student', age > 13.
     //            'student_y', age < 13. 'unknown', signed out users
```
Comment (breville) [inline apps/src/dance/DanceVisualizationColumn.jsx]: Should update comment here, especially since `"student_y"` no longer counts.

**Review by breville — APPROVED** [review]

**Comment by breville** [comment]
It sounds as though a longer-term change could be to update `user_app_options` to include this information.  It's an async call but level load waits for it to complete before proceeding.

**Comment by epeach** [comment]
Future work to tidy up and use `user_app_options` tracked in Jira: https://codedotorg.atlassian.net/browse/STAR-2177

**Pull request merged**